# U23AI009 ABSS ASSIGNMENT 01

## Loading the Data

In [38]:
import pandas as pd
df = pd.read_csv("biomet_data.csv", sep="\t", header = None)

In [39]:
df = df.T

## Inspecting & Preprocessing the Data

In [40]:
df.head()

,0,1,2,3,4,5,6,7,8,9,...,134,135,136,137,138,139,140,141,142,143
0,307.02177,364.09315,239.10673,206.98782,192.14200,339.33542,378.60285,331.73323,312.88632,243.36447,...,211.22383,244.01569,152.51256,151.69437,154.11718,149.06571,136.48869,138.14310,171.01205,161.21722
1,296.77448,348.58811,238.26485,241.55478,216.70194,292.16018,340.54587,311.61086,329.02780,229.21348,...,239.05821,258.53250,178.20929,148.15210,160.86317,153.16811,154.24074,160.82182,186.12079,171.02661
2,213.03365,331.81183,289.61825,281.39093,245.03126,337.09025,412.49380,290.51800,336.54955,215.48637,...,257.41436,302.24734,241.98635,175.44978,188.84960,168.49079,169.17194,169.52535,175.06606,178.98578
3,195.24468,289.80958,173.42357,219.13554,216.16553,272.94746,340.95611,269.46268,277.07710,234.22016,...,186.41690,241.72619,237.73706,171.83441,189.91721,147.21565,157.72283,177.64754,186.51449,182.68898
4,229.44699,327.11110,281.41246,252.41240,220.47665,380.33309,391.79664,313.81984,278.82955,256.83308,...,250.04025,280.92582,223.92805,209.27869,199.58926,150.32774,163.49850,175.31746,195.11172,197.30582


In [41]:
df.shape

(1001, 144)

In [42]:
df.info()

<class 'pandas.DataFrame'>
Index: 1001 entries, 0 to 1000
Columns: 144 entries, 0 to 143
dtypes: float64(144)
memory usage: 1.1 MB


In [43]:
df.columns.tolist()

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143]

In [44]:
df.tail()

,0,1,2,3,4,5,6,7,8,9,...,134,135,136,137,138,139,140,141,142,143
996,105.609920,141.09794,128.19904,144.03013,115.97421,135.019210,154.26094,145.18276,128.94643,114.03952,...,164.00669,191.19953,90.124298,90.717030,95.808088,113.992550,93.222355,88.181215,83.995578,90.885014
997,143.893340,130.44142,180.96623,182.73806,145.43928,150.112080,158.65187,150.25479,131.41593,119.12209,...,151.00173,173.52576,87.045179,90.210282,83.124460,88.950190,98.297011,87.139533,103.329210,114.102830
998,90.366463,114.07967,115.41041,124.57349,114.54065,85.809566,137.17225,151.72680,147.61408,113.95927,...,101.03894,153.10902,106.241430,81.078209,67.918562,79.303895,82.817479,78.115101,74.466986,79.656619
999,133.564390,131.75557,216.64343,222.68632,125.31532,173.402570,168.71354,145.17963,124.70133,103.98002,...,149.12697,193.49023,107.717260,90.621247,82.618968,102.377700,100.869440,99.349719,89.635974,94.542155
1000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [45]:
df.isnull().sum().sum()

np.int64(144)

In [46]:
df = df.dropna().reset_index(drop=True)

In [47]:
df.shape

(1000, 144)

In [48]:
df.isnull().sum().sum()

np.int64(0)

## Creating User IDs
We assume every the data is arranged in an ordered format, so like first 10 images belong to first person, next 10 to the next and so on

In [49]:
import numpy as np

# User IDs: 0,0,...,0,1,1,...,99
user_ids = np.repeat(np.arange(100), 10)

# Sample numbers: 0-9 repeated for each user
sample_ids = np.tile(np.arange(10), 100)

# Add them to the dataframe
df["user_id"] = user_ids
df["sample_id"] = sample_ids

In [50]:
df.head(12)

,0,1,2,3,4,5,6,7,8,9,...,136,137,138,139,140,141,142,143,user_id,sample_id
0,307.02177,364.09315,239.10673,206.98782,192.14200,339.33542,378.60285,331.73323,312.88632,243.36447,...,152.51256,151.69437,154.11718,149.06571,136.48869,138.143100,171.012050,161.217220,0,0
1,296.77448,348.58811,238.26485,241.55478,216.70194,292.16018,340.54587,311.61086,329.02780,229.21348,...,178.20929,148.15210,160.86317,153.16811,154.24074,160.821820,186.120790,171.026610,0,1
2,213.03365,331.81183,289.61825,281.39093,245.03126,337.09025,412.49380,290.51800,336.54955,215.48637,...,241.98635,175.44978,188.84960,168.49079,169.17194,169.525350,175.066060,178.985780,0,2
3,195.24468,289.80958,173.42357,219.13554,216.16553,272.94746,340.95611,269.46268,277.07710,234.22016,...,237.73706,171.83441,189.91721,147.21565,157.72283,177.647540,186.514490,182.688980,0,3
4,229.44699,327.11110,281.41246,252.41240,220.47665,380.33309,391.79664,313.81984,278.82955,256.83308,...,223.92805,209.27869,199.58926,150.32774,163.49850,175.317460,195.111720,197.305820,0,4
5,249.84867,327.28334,251.26869,241.61543,208.92883,299.60124,374.50100,264.85191,307.01580,227.07202,...,208.16375,183.18159,182.47739,148.53089,180.51355,191.845640,173.471090,162.911040,0,5
6,293.53549,352.94458,267.15606,245.33423,209.83645,399.31781,423.21616,342.63097,306.75948,286.87035,...,217.00418,191.81254,205.77881,167.59427,167.01982,165.720620,179.197630,163.613290,0,6
7,170.68021,373.86327,338.42246,290.04645,299.29639,285.52741,368.38783,306.78554,434.34420,376.77756,...,272.88443,190.34437,197.63227,179.95964,174.11881,179.101570,193.062170,180.613550,0,7
8,276.23258,305.48485,232.21276,231.54125,201.96145,401.11234,343.82913,313.59059,275.15654,268.53329,...,157.15857,184.67067,156.34139,145.94157,146.05747,144.168590,156.625020,144.396880,0,8
9,289.47556,309.39232,233.23617,223.89134,204.51265,263.85440,301.36457,278.97450,294.19791,230.04347,...,178.78660,166.59167,168.30024,147.71453,158.52259,154.726410,159.208410,151.510030,0,9


## Dividing into Test & Train Datasets

In [51]:
train_df = df[df["sample_id"] < 5].reset_index(drop=True)
test_df  = df[df["sample_id"] >= 5].reset_index(drop=True)

In [52]:
print(train_df.shape)
print(test_df.shape)

(500, 146)
(500, 146)


## Calculating Euclidean Distance and Cosine Similarities

In [53]:
import numpy as np
from sklearn.metrics.pairwise import euclidean_distances, cosine_similarity

# Feature vectors only
X_train = train_df.iloc[:, :144].to_numpy(dtype=float)
X_test = test_df.iloc[:, :144].to_numpy(dtype=float)

# Corresponding user IDs
train_users = train_df["user_id"].to_numpy()
test_users = test_df["user_id"].to_numpy()

In [54]:
# Euclidean distance matrix
euclidean_matrix = euclidean_distances(X_test, X_train)

# Cosine similarity matrix
cosine_matrix = cosine_similarity(X_test, X_train)

In [55]:
print(euclidean_matrix.shape)
print(cosine_matrix.shape)

print(euclidean_matrix[:5, :5])
print(cosine_matrix[:5, :5])

(500, 500)
(500, 500)
[[673.07529132 470.14646736 278.7817345  835.73219416 395.14076758]
 [385.68460716 378.91824796 554.38388111 751.52867245 341.78281307]
 [669.01253465 692.82840296 767.02209317 860.10518451 667.74580444]
 [447.93814079 352.35361258 546.66428799 869.08293772 379.31120634]
 [518.21536754 320.59192197 390.35983013 830.26221663 384.59956808]]
[[0.97677896 0.98860337 0.9960419  0.9654637  0.99192659]
 [0.99637617 0.99457603 0.98535176 0.97282639 0.99507851]
 [0.98379282 0.97984171 0.97324569 0.96574965 0.98052497]
 [0.99075847 0.99383793 0.98476102 0.96265909 0.99261535]
 [0.98480099 0.99455183 0.99310365 0.96660497 0.99277568]]


## Getting the Genuine and Impostor Distributions

In [56]:
# Lists to store scores
genuine_euclidean = []
impostor_euclidean = []

genuine_cosine = []
impostor_cosine = []

# Separate the scores
for i in range(len(test_users)):
    for j in range(len(train_users)):
        if test_users[i] == train_users[j]:
            genuine_euclidean.append(euclidean_matrix[i, j])
            genuine_cosine.append(cosine_matrix[i, j])
        else:
            impostor_euclidean.append(euclidean_matrix[i, j])
            impostor_cosine.append(cosine_matrix[i, j])

In [57]:
print("Genuine Euclidean:", len(genuine_euclidean))
print("Impostor Euclidean:", len(impostor_euclidean))

print("Genuine Cosine:", len(genuine_cosine))
print("Impostor Cosine:", len(impostor_cosine))

Genuine Euclidean: 2500
Impostor Euclidean: 247500
Genuine Cosine: 2500
Impostor Cosine: 247500


In [58]:
import plotly.express as px

In [59]:
import plotly.io as pio
pio.renderers.default = "browser"

In [60]:
import plotly.express as px
import pandas as pd

euclidean_df = pd.DataFrame({
    "Score": genuine_euclidean + impostor_euclidean,
    "Type": (["Genuine"] * len(genuine_euclidean)) +
            (["Impostor"] * len(impostor_euclidean))
})

fig = px.histogram(
    euclidean_df,
    x="Score",
    color="Type",
    nbins=50,
    barmode="overlay",
    histnorm="probability density",
    opacity=0.6,
    title="Genuine vs Impostor Distribution (Euclidean Distance)"
)

fig.update_layout(
    xaxis_title="Euclidean Distance",
    yaxis_title="Frequency"
)

fig.show()

In [61]:
cosine_df = pd.DataFrame({
    "Score": genuine_cosine + impostor_cosine,
    "Type": (["Genuine"] * len(genuine_cosine)) +
            (["Impostor"] * len(impostor_cosine))
})

fig = px.histogram(
    cosine_df,
    x="Score",
    color="Type",
    nbins=50,
    barmode="overlay",
    histnorm="probability density",
    opacity=0.6,
    title="Genuine vs Impostor Distribution (Cosine Similarity)"
)

fig.update_layout(
    xaxis_title="Cosine Similarity",
    yaxis_title="Frequency"
)

fig.show()

In [62]:
import numpy as np

print("=== Euclidean ===")
print(f"Genuine : mean={np.mean(genuine_euclidean):.3f}, std={np.std(genuine_euclidean):.3f}")
print(f"Impostor: mean={np.mean(impostor_euclidean):.3f}, std={np.std(impostor_euclidean):.3f}")

print()

print("=== Cosine ===")
print(f"Genuine : mean={np.mean(genuine_cosine):.3f}, std={np.std(genuine_cosine):.3f}")
print(f"Impostor: mean={np.mean(impostor_cosine):.3f}, std={np.std(impostor_cosine):.3f}")

=== Euclidean ===
Genuine : mean=346.858, std=173.229
Impostor: mean=732.048, std=312.219

=== Cosine ===
Genuine : mean=0.986, std=0.013
Impostor: mean=0.957, std=0.019


## Calculating FAR & FRR

### Euclidean FAR & FRR
Calculating FAR & FRR for Euclidean distance with multiple different thresholds.

In [63]:
import numpy as np

euclidean_thresholds = np.linspace(
    min(genuine_euclidean + impostor_euclidean),
    max(genuine_euclidean + impostor_euclidean),
    200
)

In [64]:
far_euclidean = []
frr_euclidean = []

genuine = np.array(genuine_euclidean)
impostor = np.array(impostor_euclidean)

for threshold in euclidean_thresholds:

    # Genuine rejected
    FRR = np.mean(genuine > threshold)

    # Impostor accepted
    FAR = np.mean(impostor <= threshold)

    far_euclidean.append(FAR)
    frr_euclidean.append(FRR)

In [65]:
far_frr_df = pd.DataFrame({
    "Threshold": np.concatenate([euclidean_thresholds, euclidean_thresholds]),
    "Rate": np.concatenate([far_euclidean, frr_euclidean]),
    "Metric": ["FAR"] * len(euclidean_thresholds) +
              ["FRR"] * len(euclidean_thresholds)
})

fig = px.line(
    far_frr_df,
    x="Threshold",
    y="Rate",
    color="Metric",
    title="FAR and FRR vs Threshold (Euclidean Distance)"
)

fig.update_layout(
    xaxis_title="Euclidean Distance Threshold",
    yaxis_title="Error Rate"
)

fig.show()

### Cosine FAR & FRR

In [66]:
cosine_thresholds = np.linspace(
    min(genuine_cosine + impostor_cosine),
    max(genuine_cosine + impostor_cosine),
    200
)

In [67]:
far_cosine = []
frr_cosine = []

genuine = np.array(genuine_cosine)
impostor = np.array(impostor_cosine)

for threshold in cosine_thresholds:

    # Genuine rejected
    FRR = np.mean(genuine < threshold)

    # Impostor accepted
    FAR = np.mean(impostor >= threshold)

    far_cosine.append(FAR)
    frr_cosine.append(FRR)

In [68]:
far_frr_df = pd.DataFrame({
    "Threshold": np.concatenate([cosine_thresholds, cosine_thresholds]),
    "Rate": np.concatenate([far_cosine, frr_cosine]),
    "Metric": ["FAR"] * len(cosine_thresholds) +
              ["FRR"] * len(cosine_thresholds)
})

fig = px.line(
    far_frr_df,
    x="Threshold",
    y="Rate",
    color="Metric",
    title="FAR and FRR vs Threshold (Cosine Similarity)"
)

fig.update_layout(
    xaxis_title="Cosine Similarity Threshold",
    yaxis_title="Error Rate"
)

fig.show()

## Receiver Originality Characteristics
Comparing ROC with both euclidean distance and cosine similarity

In [69]:
roc_compare = pd.DataFrame({
    "FAR": np.concatenate([far_euclidean, far_cosine]),
    "GAR": np.concatenate([
        1 - np.array(frr_euclidean),
        1 - np.array(frr_cosine)
    ]),
    "Method": (["Euclidean"] * len(far_euclidean)) +
              (["Cosine"] * len(far_cosine))
})

fig = px.line(
    roc_compare,
    x="FAR",
    y="GAR",
    color="Method",
    title="ROC Curve Comparison"
)

fig.update_layout(
    xaxis_title="False Acceptance Rate (FAR)",
    yaxis_title="Genuine Acceptance Rate (GAR)"
)

fig.show()

## Equal Error Rate
The point where FAR = FRR. The lower the EER, the better the system.

### Euclidean EER

In [70]:
# Convert to NumPy arrays
far_e = np.array(far_euclidean)
frr_e = np.array(frr_euclidean)

# Index where FAR and FRR are closest
eer_idx_e = np.argmin(np.abs(far_e - frr_e))

eer_threshold_e = euclidean_thresholds[eer_idx_e]
eer_e = (far_e[eer_idx_e] + frr_e[eer_idx_e]) / 2

print("=== Euclidean ===")
print(f"EER Threshold : {eer_threshold_e:.3f}")
print(f"EER           : {eer_e:.4f}")

=== Euclidean ===
EER Threshold : 476.375
EER           : 0.1558


### Cosine EER

In [71]:
far_c = np.array(far_cosine)
frr_c = np.array(frr_cosine)

eer_idx_c = np.argmin(np.abs(far_c - frr_c))

eer_threshold_c = cosine_thresholds[eer_idx_c]
eer_c = (far_c[eer_idx_c] + frr_c[eer_idx_c]) / 2

print("=== Cosine ===")
print(f"EER Threshold : {eer_threshold_c:.5f}")
print(f"EER           : {eer_c:.4f}")

=== Cosine ===
EER Threshold : 0.97464
EER           : 0.1146


Using Cosine similarity gives us an EER of 0.1146 compared to 0.1558 with Euclidean Distance. Therefore cosine similarity is an objectively superior method for a biometric identification system.

## Decidability Index
The decidability Index tells us how "decidable" the data is, by comparing the means and standard deviations of the genuine and impostor scores. The higher the decidability the better the system

### Euclidean Decidability

In [72]:
mu_g = np.mean(genuine_euclidean)
mu_i = np.mean(impostor_euclidean)

sigma_g = np.std(genuine_euclidean)
sigma_i = np.std(impostor_euclidean)

d_euclidean = abs(mu_g - mu_i) / np.sqrt((sigma_g**2 + sigma_i**2) / 2)

print("Euclidean Decidability Index:", d_euclidean)

Euclidean Decidability Index: 1.5256452863661527


### Cosine Decidability

In [73]:
mu_g = np.mean(genuine_cosine)
mu_i = np.mean(impostor_cosine)

sigma_g = np.std(genuine_cosine)
sigma_i = np.std(impostor_cosine)

d_cosine = abs(mu_g - mu_i) / np.sqrt((sigma_g**2 + sigma_i**2) / 2)

print("Cosine Decidability Index:", d_cosine)

Cosine Decidability Index: 1.7581884656912483


As expected, Cosine Similarity has a higher decidability index.

Decidability Index is inversely proportional to Equal Error Rate